README.md
# Discriminant Non-Negative Matrix Factorization

[![Travis-CI Build Status](https://travis-ci.org/zhilongjia/DNMF.png?branch=master)](https://travis-ci.org/zhilongjia/DNMF) [![CRAN Version](https://www.r-pkg.org/badges/version/DNMF)](https://cran.r-project.org/package=DNMF)

Discriminant Non-Negative Matrix Factorization is to extend the Non-negative Matrix Factorization algorithm in order to extract features that enforce not only the spatial locality, but also the separability between classes in a discriminant manner. Two kinds of Discriminant Non-Negative Matrix Factorization were implemented so far.

![Type 1 motiflogo](figure/DNMF.png)


Installation:

	devtools::install_github("zhilongjia/DNMF")

Citation:

Jia Z, Zhang X, Guan N, Bo X, Barnes MR, Luo Z (2015) [Gene Ranking of RNA-Seq Data via Discriminant Non-Negative Matrix Factorization](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0137782). PLoS ONE 10(9): e0137782. doi:10.1371/journal.pone.0137782

Reference: 
+ Zafeiriou, Stefanos, et al. [*Exploiting discriminant information in nonnegative matrix factorization with application to frontal face verification.*](https://pubmed.ncbi.nlm.nih.gov/16722172/) Neural Networks, IEEE Transactions on 17.3 (2006): 683-695.
+ Kim, Bo-Kyeong, and Soo-Young Lee. [*Spectral Feature Extraction Using dNMF for Emotion Recognition in Vowel Sounds.*](https://link.springer.com/chapter/10.1007/978-3-642-42051-1_59) Neural Information Processing. Springer Berlin Heidelberg, 2013.
+ Lee, Soo-Young, Hyun-Ah Song, and Shun-ichi Amari. [*A new discriminant NMF algorithm and its application to the extraction of subtle emotional differences in speech.*](https://link.springer.com/article/10.1007/s11571-012-9213-1#page-1) Cognitive neurodynamics 6.6 (2012): 525-535.



NEWS.md
# DNMF version 1.4 (2022-05-09)

* debug NOTEs (no visible binding for global variable runif...).

# DNMF version 1.3 (2015-06-09)

* debug NOTEs (no visible binding for global variable).

# DNMF version 1.2 (2015-06-09)

* revise typo in manual

# DNMF version 1.1 (2015-05-13)

* add nDNMF and NMFpval functions.

* init H reasonably.

* options to deal with multi-class for rnk

# DNMF version 1.0 (2015-02-16)

* Submit DNMF to CRAN.


R/rnk.R
#' write rnk to a file from matrix W.
#' 
#' write a rnk file from matrix W in a returned object of function \code{DNMF}.
#' The rnk format is referred \href{https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats}{RNK}
#' 
#' @param object a returned object of function \code{DNMF}
#' @param fn the output filename. Default is "./tmp.rnk"
#' @param type type o2m (Default) or o2o. to compare with multi sample labels. 
#' o2m means one Vs others, while o2o means one Vs another one.
#' @importFrom utils write.table
#' @export
#' @examples
#' \dontrun{
#' rnk(dnmf_result, fn="tmp.rnk")
#' }
rnk <- function (object, fn="./tmp.rnk", type="o2m"){
    
    type <- match.arg(type, c("o2m", "o2o"))
    r <- object$r
    if (r==2) {
        write.table(object$rnk, fn, sep="\t", quote=FALSE, col.names=FALSE, row.names=TRUE)
    } else if (type=="o2o"){
        for (i in 1:(r-1)) {
            for (j in (i+1):r) {
                write.table(object$W[,j]-object$W[,i], paste0(dirname(fn), "/", j, "_", i, "_", basename(fn)), sep="\t", quote=FALSE, col.names=FALSE, row.names=TRUE)
            }
        }
    } else if (type=="o2m") {
        for (i in 1:r) {
            write.table(object$W[,i]-rowMeans(object$W[,-i]), paste0(dirname(fn), "/", i, "_", basename(fn)), sep="\t", quote=FALSE, col.names=FALSE, row.names=TRUE)
        }
    }
}




R/dNMF.R
#' discrimant Non-Negative Matrix Factorization (dNMF)
#' 
#'  The dNMF algorithm with the additional Fisher criterion on the cost 
#'  function of conventional NMF was designed to increase class-related
#'  discriminating power.
#'  
#'  This algorithm is based on articles.
#'  \enumerate{
#'  \item Kim, Bo-Kyeong, and Soo-Young Lee. "Spectral Feature Extraction Using dNMF for Emotion Recognition in Vowel Sounds." Neural Information Processing. Springer Berlin Heidelberg, 2013.
#'  \item Lee, Soo-Young, Hyun-Ah Song, and Shun-ichi Amari. "A new discriminant NMF algorithm and its application to the extraction of subtle emotional differences in speech." Cognitive neurodynamics 6.6 (2012): 525-535.
#'  }
#' @param dat a matrix with gene in row and sample in column
#' @param trainlabel the label of sample, like c(1,1,2,2,2)
#' @param r the dimension of expected reduction dimension, with the default value 2
#' @param lambada a relative weighting factor for the discriminant. Default 0.1
#' @param maxIter the maximum iteration of update rules, with the default value 1000
#' @param tol the toleration of coverange, with the default value 1e-7
#' @param log log2 data. Default is TRUE.
#' @param verbose TRUE
#' @author Zhilong Jia and Xiang Zhang
#' @export
#' @import Matrix
#' @examples
#' dat = rbind(matrix(c(rep(3, 16), rep(8, 24)), ncol=5), 
#' matrix(c(rep(5, 16), rep(5, 24)), ncol=5), 
#' matrix(c(rep(18, 16), rep(7, 24)), ncol=5)) + 
#' matrix(runif(120,-1,1), ncol=5)
#' trainlabel <- c(1,1,2,2,2)
#' 
#' res <- dNMF(dat, trainlabel, r=2, lambada = 0.1)
#' res$H
#' res$W[,2]- res$W[,1]
#' 

dNMF <- function(dat, trainlabel, r=2, lambada=0.1, maxIter=1000, tol=1e-7, log=TRUE, plotit=FALSE, verbose=FALSE){
    
    eps = .Machine$double.eps
    dat <- as.matrix(dat)
    dat[which(dat==0)] <- 1
    
    if (log){
        dat <- log2(dat + 2) 
    }
    
    nFea = nrow(dat); nSmp = ncol(dat)
    
    # init the H0 and W0 matrix 
    # The 1st row of H is down-regualted genes, 
    # while the 2nd row of H is up-regualted genes)
    H = matrix(runif(r*nSmp, eps), r, nSmp)
    for (i in 1:r){
        H[i,which(trainlabel==names(table(trainlabel))[i])] = H[i,which(trainlabel==names(table(trainlabel))[i])] + sum(H)
    }
    H = pmax(H,0)
    W = matrix(runif(nFea*r, eps), nFea, r)
    # W = W/colSums(W)
    W = pmax(W, eps)
    b = pmax(abs(W %*% H), eps)
    
    obj0 = -10
    lambada = lambada * nFea/r
    
    #averaing matrix Ma over all samples and Mc for wach class
    Ma = matrix(1/nSmp, nrow=nSmp, ncol=nSmp)
    Mc <- list()
    for (i in unique(trainlabel)){
        Mc[[i]] = matrix(1/table(trainlabel)[i], table(trainlabel)[i], table(trainlabel)[i])
    }
    Mc = as.matrix(Matrix::bdiag(Mc))
    
    Hclass = H
    final = Inf
    count = 1
    obj_stack =  vector (mode="numeric", length = maxIter)
    
    while (final > tol && count <= maxIter) {
        
        # update W and H
        W = W * (dat %*% t(H) / (W %*% (H %*% t(H)) + eps) )
        H = H * ( (t(W) %*% dat + lambada * H %*% Mc) / (t(W) %*% W %*% H + lambada * H %*% Ma + eps) )
        if (verbose) print (count)
        
        # H normalization (Formula 11)
        H = as.matrix(H)
        H = H / (rowSums(H)/nSmp + eps)
        
        # Convergence
        b <- pmax(abs( W%*%H ), eps)
        obj1 = norm(dat-b+eps, type="F")^2 - lambada * sum(diag(H %*% (Ma-Mc) %*% t(H)))
        final = abs(obj0-obj1) / (abs(obj0) + eps)
        obj0 = obj1
        if (verbose) print (final)
        obj_stack[count] = obj1
        count = count + 1
        
    }
    
    # to plot the convergence of the object function
    if (plotit){
        heatmap(H)
    }
    
    list(V=dat, W=W, H=H, trainlabel=trainlabel, count=count,
         final=final, obj_stack=obj_stack, r=r, call=match.call())
}




R/ndNMF.R
#' a new discriminant Non-Negative Matrix Factorization (dNMF)
#' 
#'  The ndNMF algorithm with the additional Fisher criterion on the cost 
#'  function of conventional NMF was designed to increase class-related
#'  discriminating power.
#'  
#'  This algorithm is based on articles.
#'  \enumerate{
#'  \item Kim, Bo-Kyeong, and Soo-Young Lee. "Spectral Feature Extraction Using dNMF for Emotion Recognition in Vowel Sounds." Neural Information Processing. Springer Berlin Heidelberg, 2013.
#'  \item Lee, Soo-Young, Hyun-Ah Song, and Shun-ichi Amari. "A new discriminant NMF algorithm and its application to the extraction of subtle emotional differences in speech." Cognitive neurodynamics 6.6 (2012): 525-535.
#'  }
#' @param dat a matrix with gene in row and sample in column
#' @param trainlabel the label of sample, like c(1,1,2,2,2)
#' @param r the dimension of expected reduction dimension, with the default value 2
#' @param lambada a relative weighting factor for the discriminant. Default 0.1
#' @param maxIter the maximum iteration of update rules, with the default value 1000
#' @param tol the toleration of coverange, with the default value 1e-7
#' @param log log2 data. Default is TRUE.
#' @param plotit whether plot H (V=WH). Default: FALSE.
#' @param verbose TRUE
#' @param ... to gplots::heatmap.2
#' @author Zhilong Jia and Xiang Zhang
#' @export
#' @import Matrix
#' @importFrom stats runif
#' @examples
#' dat <- rbind(matrix(c(rep(3, 16), rep(8, 24)), ncol=5), 
#' matrix(c(rep(5, 16), rep(5, 24)), ncol=5), 
#' matrix(c(rep(18, 16), rep(7, 24)), ncol=5)) + 
#' matrix(runif(120,-1,1), ncol=5)
#' trainlabel <- c(1,1,2,2,2)
#' 
#' res <- ndNMF(dat, trainlabel, r=2, lambada = 0.1)
#' res$H
#' res$rnk
#' 

ndNMF <- function(dat, trainlabel, r=2, lambada=0.1, maxIter=1000, tol=1e-7, log=TRUE, plotit=FALSE, verbose=FALSE, ...){
    
    eps = .Machine$double.eps
    dat <- as.matrix(dat)
    dat[which(dat==0)] <- 1
    
    if (log){
        dat <- log2(dat + 2) 
    }
    
    nFea = nrow(dat); nSmp = ncol(dat)
    
    # init the H0 and W0 matrix 
    # The 1st row of H is down-regualted genes, 
    # while the 2nd row of H is up-regualted genes)
    H = matrix(runif(r*nSmp, eps), r, nSmp)
    for (i in 1:r){
        H[i,which(trainlabel==names(table(trainlabel))[i])] = H[i,which(trainlabel==names(table(trainlabel))[i])] + sum(H)
    }
    H = pmax(H,0)
    W = matrix(runif(nFea*r, eps), nFea, r)
    # W = W/colSums(W)
    W = pmax(W, eps)
    b = pmax(abs(W %*% H), eps)
    
    obj0 = -10
    lambada = lambada * nFea/r
    
    #averaing matrix Ma over all samples and Mc for wach class
    Ma = matrix(1/nSmp, nrow=nSmp, ncol=nSmp)
    Mc <- list()
    for (i in unique(trainlabel)){
        Mc[[i]] = matrix(1/table(trainlabel)[i], table(trainlabel)[i], table(trainlabel)[i])
    }
    Mc = as.matrix(Matrix::bdiag(Mc))
    
    Hclass = H
    final = Inf
    count = 1
    obj_stack =  vector (mode="numeric", length = maxIter)
    
    while (final > tol && count <= maxIter) {
        
        # update W and H
        W = W * (dat %*% t(H) / (W %*% (H %*% t(H)) + eps) )
        H = H * ( (t(W) %*% dat + lambada * H %*% Mc) / (t(W) %*% W %*% H + lambada * H %*% Ma + eps) )
        if (verbose) print (count)
        
        # H normalization (Formula 11)
        H = as.matrix(H)
        H = H / (rowSums(H)/nSmp + eps)
        
        # Convergence
        b <- pmax(abs( W%*%H ), eps)
        obj1 = norm(dat-b+eps, type="F")^2 - lambada * sum(diag(H %*% (Ma-Mc) %*% t(H)))
        final = abs(obj0-obj1) / (abs(obj0) + eps)
        obj0 = obj1
        if (verbose) print (final)
        obj_stack[count] = obj1
        count = count + 1
        
    }
    
    # to plot H
    if (plotit){
        gplots::heatmap.2(H, scale="row", trace="none", density.info="none", keysize=1, cexCol=0.8, srtCol=30, ...)
    }
    
    list(V=dat, W=W, H=H, rnk=W[,2]-W[,1], trainlabel=trainlabel, count=count,
         final=final, obj_stack=obj_stack, r=r, call=match.call())
}




R/NMFpval.R
#' P value for discriminant Non-Negative Matrix Factorization
#' 
#' Estimate the significance of differentially expressed genes in parallel.
#' 
#' P value is caculated based on aatricle, Wang, Hong-Qiang, Chun-Hou Zheng, and Xing-Ming Zhao. 
#' "jNMFMA: a joint non-negative matrix factorization meta-analysis of
#'  transcriptomics data." Bioinformatics (2014): btu679.
#' 
#' @param nmf_res result from DNMF or dNMF
#' @param np number of permutations
#' @param ncores cores used. Default is all the availiable cores
#' @param fdr false discovery rate. Default is FALSE
#' @param top only include top ranked genes. Default is 1000
#' @param verbose verbose
#' @return a matrix with columns rnk, p (and fdr)
#' @export
#' @import parallel
#' @importFrom stats p.adjust
#' @import foreach
#' @import doParallel
#' @author Zhilong Jia
#' @examples
#' dat <- rbind(matrix(c(rep(3, 16), rep(8, 24)), ncol=5), 
#' matrix(c(rep(5, 16), rep(5, 24)), ncol=5), 
#' matrix(c(rep(18, 16), rep(7, 24)), ncol=5)) + 
#' matrix(runif(120,-1,1), ncol=5)
#' trainlabel <- c(1,1,2,2,2)
#' 
#' nmf_res <- ndNMF(dat, trainlabel, r=2, lambada = 0.1)
#' pMat <- NMFpval(nmf_res, np=10, ncores=2, top=4)
#' 
NMFpval <- function (nmf_res, np=100, ncores=parallel::detectCores(), fdr=FALSE, top=1000, verbose=FALSE) {
    
    W <- nmf_res$W
    rownames(W) <- NULL
    rnk0 <- nmf_res$rnk
    rnk <- rnk0[which(abs(rnk0)>sort(abs(rnk0), decreasing=TRUE)[top] )]
    
    
    cl <- parallel::makeCluster(ncores)
    doParallel::registerDoParallel(cl)
    if (verbose) {print(paste("getDoParWorkers:", foreach::getDoParWorkers()))}
    strt<-Sys.time()
    
    j <- NULL
    matP <- foreach::foreach(1:np, .combine='cbind') %:%
        foreach::foreach(j=rnk, .combine='c') %dopar% {
            length(which(abs(j)< abs(sample(W[,2], length(rnk0)) - W[,1])))
        }
    print(Sys.time()-strt)
    parallel::stopCluster(cl)
    
    p <- rowMeans(matP)/ length(rnk0)
    if (fdr) {
        fdr <- p.adjust(p, method="fdr")
        dat <- cbind(rnk, p, fdr)
    } else {
        dat <- cbind(rnk, p)
    }
    
    return (dat)
}


R/DNMF.R
#' Discriminant Non-Negative Matrix Factorization.
#'
#' Discriminant Non-Negative Matrix Factorization, DNMF, is to extend the Non-negative Matrix Factorization algorithm in 
#' order to extract features that enforce not only the spatial locality, but
#'  also the separability between classes in a discriminant manner.
#' 
#' The main algorithm is based on 
#' \href{https://pubmed.ncbi.nlm.nih.gov/16722172/}{Zafeiriou, S., et al. 
#' (2006) Exploiting discriminant information in 
#' nonnegative matrix factorization with application to frontal face 
#' verification, IEEE transactions on neural networks, 17, 683-695}, 
#' with some \strong{CORRECTIONs}. 
#'
#' @param data a matrix, like expression profilings of some samples. the columns are samples and the rows are gene's expression.
#' @param trainlabel a numeric vector of sample type of all the samples, this vector should ONLY contain 1 and 2 so far and length of it should equal the column (sample) size of data.
#' @param r the dimension of expected reduction dimension, with the default value 2.
#' @param gamma the tradeoff value for the within scatter matrix, with the default value 0.1.  
#' @param delta the tradeoff value for the between scatter matrix, with the default value 1e-4.
#' @param maxIter the maximum iteration of update rules, with the default value 1000.
#' @param log log2 data. Default is TRUE.
#' @param tol the toleration of coverange, with the default value 1e-7.
#' @param plotit whether plot H (V=WH). Default: FALSE.
#' @param checkH whether or not check H. Default: TRUE. This parameter aims to 
#' check whether or not the H safisfy the discriminant metagenes. Usually, this
#' should be TRUE.
#' @param ... to gplots::heatmap.2
#' @import gplots
#' @importFrom stats runif
#' @author Zhilong Jia and Xiang Zhang
#' @export
#' @examples
#' dat <- rbind(matrix(c(rep(3, 16), rep(8, 24)), ncol=5), 
#' matrix(c(rep(5, 16), rep(5, 24)), ncol=5), 
#' matrix(c(rep(18, 16), rep(7, 24)), ncol=5)) + 
#' matrix(runif(120,-1,1), ncol=5)
#' trainlabel <- c(1,1,2,2,2)
#' 
#' DNMF_result <- DNMF(dat, trainlabel, r=2)
#' 
#' 
#' \dontrun{
#' # Gene ranking. dat is the raw read count maatrix with sample in column.
#' 
#' #normalising dat
#' Sizefactors <- DESeq::estimateSizeFactorsForMatrix(dat)
#' dat = sweep(dat, 2, Sizefactors, `/`)
#' 
#' res <- DNMF(dat, trainlabel, r=2)
#' rnk <- res$rnk
#' 
#' #The end of gene ranking exmaples
#' 
#' #Other exmaples
#' DNMF_result <- DNMF(dat, trainlabel, r=2, gamma=0.1, delta=0.0001, plotit=TRUE)
#' }
#' 

DNMF <- function(data,trainlabel, r=2, gamma=0.1, delta=0.0001, maxIter=1000, 
                 tol=1e-7, log=TRUE, plotit=FALSE, checkH=TRUE, ...) {
	
    data <- as.matrix(data)
    data[which(data==0)] <- 1
    
    if (log){
        data <- log2(data + 2) 
    }
    nFea = nrow(data); nSmp = ncol(data)
    eps = .Machine$double.eps
    
    #init the H0 and W0 matrix
    H = matrix(runif(r*nSmp, eps), r, nSmp)
    # The 1st row of H is down-regualted genes, 
    # while the 2nd row of H is up-regualted genes)
    for (i in 1:r){
        H[i,which(trainlabel==names(table(trainlabel))[i])] = H[i,which(trainlabel==names(table(trainlabel))[i])] + sum(H)
    }
    H = pmax(H,0)
    W = matrix(runif(nFea*r, eps), nFea, r)
    W = W/colSums(W)

    #calculate KL divergence of two matrix
    b = pmax(abs(W %*% H), eps)
    obj0 = sum(data*log((data+eps)/(b-eps))-data+b)
    obj1 = obj0 
    
    ##########################
    E = matrix(1, nFea, nSmp)
    # N is just 1/Nr in paper, the weighted matrix
    SmpCount_withinClass = vector(mode="numeric", length(trainlabel))
    SmpCount_withinClass = as.vector(table(trainlabel)[trainlabel])
    N = matrix(1/SmpCount_withinClass, r, nSmp, byrow=T)     

    final = Inf
    count = 1
    Hclass = H  
    obj_stack =  vector (mode="numeric", length = maxIter)
    
while (final > tol && count <= maxIter) {

    #update H with the objective function includes KL divergence

    for(i in unique(trainlabel)){
        Hclass[,which(trainlabel==i)] = matrix( rep(rowSums(H[,which(trainlabel==i)]), length(which(trainlabel==i))), r, length(which(trainlabel==i)))
    }
    Hclass = Hclass - H
    Hsum = matrix(rep(rowSums(H),ncol(H)),r, ncol(H)) - H

    tmp_a = 4*gamma + 4*(delta/nSmp - (gamma+delta)*N)   #2a
    tmp_b = 1 + 2*delta/nSmp * Hsum - 2*(delta+gamma)*N * Hclass
    tmp_c = - t(W) %*% (data / ( W %*% H + eps) ) * H
    H = (sqrt(tmp_b^2 - 2*tmp_a*tmp_c) - tmp_b ) /tmp_a
    H = pmax(H, eps)  

    #######################################
    #update W
    W = (W/(E%*%t(H)))*(data/(W%*%H)) %*% t(H)
    H = diag(colSums(W)) %*% H
    W = W/ matrix(rep(colSums(W), each=nrow(W)), nrow(W), ncol(W))
    W = pmax(W,eps)
    
    obj2 = obj1
    b = pmax(abs(W%*%H),eps)
    obj1 = sum( data*log((data+eps)/(b-eps)) - data + b )
    final = abs(obj1-obj2) / abs(obj1-obj0)
    #obj_stack[count] = final
    obj_stack[count] = obj1
    count = count + 1
}
    # to plot H
    if (plotit){
        gplots::heatmap.2(H, scale="row", trace="none", density.info="none", keysize=1, cexCol=0.8, srtCol=30, ...)
    }
    

    # check H. Setting down-regulted metagene in row 1 of H, 
    # while up-regulated metagene in row 2 of H.
    l1 <- apply(H[,which(trainlabel == names(table(trainlabel))[1])], 1, mean)
    l2 <- apply(H[,which(trainlabel == names(table(trainlabel))[2])], 1, mean)
    meanH <- cbind(l1, l2)
    if (checkH){
        if (l1[1] < l2[1] && l1[2] > l2[2]) {
            H <- rbind(H[2,], H[1,])
            W1 <- cbind(W[,2], W[,1]) 
        } else if ((l1[1] < l2[1] && l1[2] < l2[2]) || (l1[1] > l2[1] && l1[2] > l2[2])) {
            stop("Failed. Run DNMF again after restart R.")
        }
    }
    
    list(V=data, W=W, H=H, rnk=W[,2]-W[,1], trainlabel=trainlabel, delta=delta, gamma=gamma, count=count,
         final=final, obj_stack=obj_stack, r=r, call=match.call())
}


